documentation/extension.tex

```diff
@@ -1 +1,98 @@
+
+Our IDE choice is \emph{Visual Studio Code}, due to its popularity and lightweight design. Conveniently, \emph{Theia}, a web-based IDE, supports VSCode extensions, therefore our plugin works with \emph{Theia} as well.
+
+\section{Standard LSP Extension}
+
+The core of the extension is an activation event which starts the plugin for VSCode.
+
+Upon activation, \emph{Language Client} and \emph{Language server}~\ref{chap:lang_server} are started as child processes of VSCode and a pipe is open for their communication. The LSP communication and its features are handled by the \emph{vscodelc} package. 
+
+To be independent of pipes, we have added an option to use TCP, which assigns a random free port for TCP communication.
+
+\section{DAP Extension}
+
+\emph{Macro Tracer}~\ref{chap:macro_tracer} is implemented using DAP, which is also supported out-of-the-box by VSCode. We dynamically assign a random free port for DAP communication during the activation as well.
+
+\section{Additions}
+
+Several features are added to the extension to simplify work with HLASM in modern editors. These additions are specific for Visual Studio Code (and Theia) and are not a part of the LSP specification.
+
+\subsection{Language Detection}
+
+The usual workflow with the extension begins with downloading HLASM source codes from mainframe. Typically, these files will not have any file extension and even if they do, they might differ across various products.
+
+To cope with this problem, there are several mechanisms that help the user to recognize the file as HLASM automatically.
+
+\subsubsection{Macro Detection}
+
+Each file starting with line \emph{MACRO} (arbitrary number of whitespace before and after) is recognized as HLASM.
+
+\subsubsection{Configuration Files Detection}
+
+Every file either defined as a program or as a part of a processor group is recognized as HLASM.
+
+\subsubsection{Wildcards}
+
+Configuration file \emph{pgm\_conf.json} contains a field \emph{alwaysRecognize}, which consists of user-defined wildcards. Every file that satisfies at least one of these wildcards is recognized as HLASM.
+
+\subsubsection{Automatic Language Detection}
+
+Whenever a file is open, its contents are scanned line by line. Whether a file is HLASM or not is determined by a ratio (\# HLASM lines)/(\# all lines).
+
+Recognition is mostly based on a pre-defined set of most used instructions. If a line correctly uses one of these instructions, it is counted as a HLASM line. Continued line of a HLASM line is also a HLASM line.
+Moreover, a HLASM line must not exceed 80 characters.
+
+Comment lines or empty lines are skipped and not counted. 
+
+We tested the detection on ~25.000 HLASM files. The best results were observed using 4/10 ratio, with 88\% true positive recognition and 95\% true negative recognition. 
```

Comment:
a kolko suborov bolo ne-hlasm? z hentoho mi vychadza ze sme to testovali iba na hlasm filoch

Reply:
pravda, to tam chyba